python/ray/experimental/workflow/workflow_storage.py

```diff
@@ -0,0 +1,268 @@
+
+        Returns:
+            The status of the step.
+        """
+        step_dir = self._steps_dir / step_id
+        if not self._storage.directory_exists(step_dir):
+            return StepInspectResult()
+
+        # does this step contains output checkpoint file?
+        if self._storage.file_exists(step_dir / STEP_OUTPUT):
+            return StepInspectResult(output_object_valid=True)
+
+        # do we know where the output comes from?
+        output_step_id = self._locate_output_step(step_dir)
+        if output_step_id is not None:
+            return StepInspectResult(output_step_id=output_step_id)
+
+        # read inputs metadata
+        try:
+            metadata = self._storage.read_json(step_dir / STEP_INPUTS_METADATA)
+            input_object_refs = metadata["object_refs"]
+            input_workflows = metadata["workflows"]
+        except (FileNotFoundError, json.JSONDecodeError):
+            input_object_refs = None
+            input_workflows = None
+        return StepInspectResult(
+            args_valid=self._storage.file_exists(step_dir / STEP_ARGS),
+            object_refs=input_object_refs,
+            workflows=input_workflows,
+            func_body_valid=self._storage.file_exists(
+                step_dir / STEP_FUNC_BODY),
+        )
+
+    def validate_workflow(self) -> None:
+        """Check if the workflow structured correctly."""
+        if not self._storage.directory_exists(self._workflow_dir):
+            raise ValueError("Cannot find the workflow job "
+                             f"'{self._workflow_dir}'")
+        if not self._storage.directory_exists(self._steps_dir):
+            raise ValueError(f"The workflow job record is invalid: {STEPS_DIR}"
+                             " not found.")
+
+    def commit_step(self, step_id: StepID, ret: Union[Workflow, Any],
+                    forward_output_to: Optional[StepID]) -> None:
+        """When a workflow step returns,
+        1. If the returned object is a workflow, this means we are a nested
+           workflow. We save the DAG of the nested workflow.
+        2. Otherwise, checkpoint the
+
+        Args:
+            step_id: The ID of the workflow step. If it is an empty string,
+                it means we are in the workflow job driver process.
+            ret: The returned object from a workflow step.
+            forward_output_to: The output should also forward to the step
+                referred by 'forward_output_to'. When resume from that step,
+                that step can directly read this output.
+        """
+        if isinstance(ret, Workflow):
+            # This workflow step returns a nested workflow.
+            assert not ret.executed
+            if not ret.skip_saving_workflow_dag:
+                for w in ret.iter_workflows_in_dag():
+                    self.write_step_inputs(w.id, w.get_inputs())
+                self.write_step_output_metadata(step_id, {"step_id": ret.id})
```

Comment:
Is this last write "committing" the entire batch of writes above?

Reply:
This part does seem confusing. I have refactored the storage module and it looks clear now.